ProcessLib/ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,299 @@
+    std::vector<GlobalVector*> const& process_solutions, Process const& process,
+    Status const status)
+{
+    int const nNodes = _solutions.size();
+    for (int node_id = 0; node_id < nNodes; ++node_id)
+    {
+        int process_id = 0;
+        int const variable_id = 0;
+        auto& solution = _solutions[node_id];
+        auto& components = solution._components;
+        for (auto process_solution_iter = process_solutions.begin();
+             process_solution_iter != process_solutions.end();
+             ++process_solution_iter)
+        {
+            auto const& per_process_variables =
+                process.getProcessVariables(process_id);
+            auto const& pv_name =
+                per_process_variables[variable_id].get().getName();
+
+            auto component = std::find_if(components.begin(), components.end(),
+                                          [&pv_name](Component const& c) {
+                                              return c.getName() == pv_name;
+                                          });
+            if (component != components.end())
+            {
+                switch (status)
+                {
+                    case Status::SettingSolutions:
+                        // Set component concentrations.
+                        component->setAmount(
+                            (*process_solution_iter)->get(node_id));
+                        break;
+                    case Status::UpdatingProcessSolutions:
+                        // Update solutions of component transport processes.
+                        (*process_solution_iter)
+                            ->set(node_id, component->getAmount());
+                        break;
+                }
+            }
+
+            if (pv_name == "H")
+            {
+                switch (status)
+                {
+                    case Status::SettingSolutions:
+                    {
+                        // Set pH value by hydrogen concentration.
+                        auto pH =
+                            -log10((*process_solution_iter)->get(node_id));
+                        solution.setpH(pH);
+                        break;
+                    }
+                    case Status::UpdatingProcessSolutions:
+                    {
+                        // Update hydrogen concentration by pH value.
+                        auto hydrogen_concentration =
+                            pow(10, -solution.getpH());
+                        (*process_solution_iter)
+                            ->set(node_id, hydrogen_concentration);
+                        break;
+                    }
+                }
+            }
+
+            ++process_id;
+        }
+    }
+}
+
+void phreeqcIO::writeInputsToFile()
+{
+    DBUG("Writing phreeqc inputs into file '%s'.", _input_file);
+    std::ofstream outFile(_input_file, std::ofstream::out);
+    outFile << *this;
+    outFile.close();
```

Comment:
No need to explicitly close the file, but must check the `outFile` state before and after writing to it.

Reply:
Added a check on the state of stream.

Reply:
> No need to explicitly close the file, but must check the `outFile` state before and **after** writing to it.
It is still not checked, whether the output was successful.

Reply:
Fixed and amended into the previous commit `7f5f68223`

.